### ISSUE TITLE
Alpha Milestone Not Far to Go!

### ISSUE THREAD

**Issue by jpcsupplies** [jpcsupplies/Economy_mod#55]
Looks like we are down to last three issues relevant to our alpha release:
Key issues:
* ~~#32 Buy from server command (basically cloning sell command but adding not removing; and dropping overflow at feet) (/buy X Y)  No "all" support is needed in this context. - needs #57 to finalize~~  done :+1: 
* ~~#35 request logic (display a prompt to accept or deny X action - may be reusable in several places) can be used when selling to players, or demanding payment from a player for services~~
* ~~#46 Sell items to other players - needs working request logic functionality~~
* ~~#49 Need range check on 0,0,0 location to force trade within 5km of middle of map (few players requested this)~~  
* ~~Generally - need to add how to use the new LCD price listing functionality into manual~~
* ~~Bug #75 blank lcds~~
* ~~Final testing and release.~~

ENH or misc or OCD nitpicks
* ~~#45 should resolve itself as part of above~~
* #53 is not critical to operation of plugin but makes it inconvenient for admins to tweak pricing or blacklist.
* #20 This can wait until next milestone - as it is dependent on #49 logic becoming functional which is needed for a bunch of other cool stuff like #44 logic
* #6 purge logic - this can wait, we have everything we need, but it is not easy to implement without faction bank balances; or rethinking "asset disposal" logic..  not likely to become urgent unless servers start getting player turnover in the thousands
* #33 this is basically moving our behavior constants to the config file. Once NPC trade is finalized, this can wait until our next milestone.

**Comment by jpcsupplies** [comment]
I've released a preview of the plugin as is now to get players impressions.

So far #59 #60 and #61 is jumping out the most in that regard.

**Comment by jpcsupplies** [comment]
Ok, looks like we are just about there for the Alpha;  how ready is player to player at the moment in your opinion?

**Comment by midspace** [comment]
Would like to test it, at least with my two accounts, but every time I go to test it, I've got other things that need attention.

**Comment by jpcsupplies** [comment]
Righto, ill have a crack at testing it on my account and my daughters,  is the current build up on your workshop testing item ?

**Comment by midspace** [comment]
Yes, it's there.

**Comment by jpcsupplies** [comment]
Sounds interesting, if painful.  Keen giving you tester credit?   I did a bit for Starforge Alpha, but mostly I just invested as a founder to help fund the game and im in the game credits now.  Didn't pay anything tho, but having my name in a game was enough to impress my kids :+1:

**Comment by jpcsupplies** [comment]
Testing results:
Scenario 1:  join server
on join i am informed timers ran out, type /sell collect to retrive items
two rifles return to inventory.  (from my last test apparently)

Scenario 2: sell using incorrect syntax (alone on server)
/sell 1 rifle lana 10
it then appeared to sell to the NPC market instead of giving an error or such

Scenario 3: sell to online player
lana then connected
and i did /sell 1 rifle 10 lana
the offer sent to lana, on typing /sell accept
i got a message 
you just sold 10 worth of xxx (0 units)
you just purchased 10 worth of xxx (0 units)
nothing appears to add to inventory

Scenario 4: deny the offer
if i sell deny i get not yet complete and offer still remains active

Scenario 5: try to sell to offline player with correct syntax
if i do /sell 1 rifle 10 lana
i get sell not yet complete error
but rifle is removed from inventory no timeout seems to occur (but i may be just impatient)

Scenario 6:  use /sell collect
2 rifles returned to my inventory

**Comment by jpcsupplies** [comment]
Did notice a bug in main game tho, after lana disconnected she still shows on connected players list

**Comment by midspace** [comment]
Updates done to #46 just now.

> Scenario 1: join server

haven't touched.

> Scenario 2: sell using incorrect syntax (alone on server)

I've updated the regex to try and fix this.

> Scenario 3: sell to online player

fixed messages sent to wrong players.
automatically adding to inventory if there is space.
If not, new message.

> Scenario 4: deny the offer
haven't touched.

> Scenario 5: try to sell to offline player with correct syntax

Locked out selling to offline players with new message.
Inventory won't be touched.

> Scenario 6: use /sell collect

Have updated some messages, though more work needs doing on the wording.

**Comment by jpcsupplies** [comment]
Sell collect seemed to be working pretty well.  Even when items disappeared without notifications; all the items were preserved, and returned when using sell collect. So from a data integrity perspective, it is working so well that nothing gets lost even when something goes wrong.

**Comment by jpcsupplies** [comment]
DS testing
scenario 1:
/sell 1 rifle 10 dragon_hunter
result success

scenario 2:
/sell all "iron ore" 0.001 dragon_hunter
result apparent success but the sale total looked a bit odd will test further


How are we looking with our 2500 metres range limit on trade zone 0,0,0 (NPC center of map)

**Comment by jpcsupplies** [comment]
Can we also get stock on hand for lcds too ? 
unlike the rest of the game, economy mod seems to be working flawlessly lol

**Comment by midspace** [comment]
> How are we looking with our 2500 metres range limit on trade zone 0,0,0 (NPC center of map)

Done.

> Can we also get stock on hand for lcds too - and perhaps not show blacklisted items (or does it do this already?).

We can. You need to have a look at how much space is on the LCD.
We're (I'm) going to need a more definitive list of settings to apply to the LCD and what each setting presents.

> Rangecheck on 0,0,0

It's in, and is stored in the Data file. So, eventually we can make it configurable via commands in game.

> we need a way to integrate changes

The /set command will need to be expanded. I've put in some of the necessary mechanisms in `MessageSet`. The parsing of /set command itself needs expanding to include the new options.

> Another issue is I cant seem to work out how to get tools category onto LCD 

I didn't create one.
I'll have a look at that now.

**Comment by midspace** [comment]
> Another issue is I cant seem to work out how to get tools category onto LCD 

Its in now. https://github.com/jpcsupplies/Economy_mod/commit/353bf894ad3aa84f111c5e759990f0003e1fe643

> and perhaps not show blacklisted items

It's in now too.

**Comment by jpcsupplies** [comment]
OK I think thats everything planned for the Alpha milestone, and then some! everything else is an ENH for the next version..   all we need to do is test and update our manual ?

**Comment by jpcsupplies** [comment]
Up for some proof reading ?  
http://steamcommunity.com/sharedfiles/filedetails/?id=481264606

**Comment by midspace** [comment]
not `/sell claim`  it's `/sell collect`
missing `/sell deny`
missing `/sell cancel`

**Comment by jpcsupplies** [comment]
Done

**Comment by jpcsupplies** [comment]
Ok, i think we nailed all our dev targets for this Milestones, lcds are all good, cant see any errors, ill just tweak the major version string and bash out an official release..   any issues left you think i should wait on?   Cause it looks all good so far.

**Comment by midspace** [comment]
something that comes to mind.
Testing with Zan last night, /pay does not appear to be range restricted.

**Comment by jpcsupplies** [comment]
Pay is like an electronic wire transfer, probably doesnt need range limits.. otherwise, how can you order a shipment of beer from earth etc :pager:

**Comment by midspace** [comment]
I haven't looked at it, but it may be ignoring for Admin.

**Comment by jpcsupplies** [comment]
I would even go so far as to say we shouldn't have a range check on payments; it would limit the usefulness too much.   Range checks on trading goods however is perfectly appropriate..
